File: qyao/213-house-robber-II/Solution.py
def house_rob_I(nums):
    if len(nums) == 0:
        return 0
    if len(nums) == 1:
        return nums[0]

    dp = [0, 0] # initialize circular buffer

    dp[0] = 0
    dp[1] = nums[0]

    for i in range(2, len(nums)+1):
        dp[i%2] = max(dp[(i-2)%2] + nums[i-1], dp[(i-1)%2])

    return dp[len(nums)%2]


class Solution(object):
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        """
        idea: use the same idea as house robber I, except this time, we only have 2 cases, either we rob the last house or not rob the first house
        assuming the solution to house robber I is f(), then choosing to rob the first house would be nums[0] + f(nums[2:-1]), and not choosing to rob the first house would be f(nums[1:])
        """
        if len(nums) == 0:
            return 0
        elif len(nums) == 1:
            return nums[0]
        
        return max(nums[0] + house_rob_I(nums[2:-1]), house_rob_I(nums[1:]) )
        

File: qyao/213-house-robber-II/Solution.java
class Solution {
  /*
   * Dynamic programming, the optimal amount for the first k houses is max(S_k-1, S_k-2 + V_k)
   * Using a circular buffer as we only need to remember k-1 and k-2
   * Difference from house robber 1: there can be only two cases, rob the first house or not
   * TODO should use a helper function specifying the start and end
   */
  public int rob(int[] nums) {
    if (nums.length == 0) {
      return 0;
    } else if (nums.length == 1) {
      return nums[0];
    } else if (nums.length == 2) {
      return Math.max(nums[0], nums[1]);
    }

    // case 1: rob the first house
    int[] circularBuffer = new int[2];
    circularBuffer[0] = nums[0];
    circularBuffer[1] = Math.max(nums[0], nums[1]);
    for (int i = 2; i < nums.length - 1; i++) {
      circularBuffer[i % 2] =
          Math.max(circularBuffer[(i - 1) % 2], circularBuffer[(i - 2) % 2] + nums[i]);
    }
    int robFirstHouse = circularBuffer[(nums.length - 2) % 2];

    // case 2: don't rob the first house
    circularBuffer[0] = 0;
    circularBuffer[1] = nums[1];
    for (int i = 2; i < nums.length; i++) {
      circularBuffer[i % 2] =
          Math.max(circularBuffer[(i - 1) % 2], circularBuffer[(i - 2) % 2] + nums[i]);
    }
    int doNotRobFirstHouse = circularBuffer[(nums.length - 1) % 2];

    return Math.max(robFirstHouse, doNotRobFirstHouse);
  }
}
